syntax = "proto3";
import "google/protobuf/timestamp.proto";

message Discount {
  float pct = 1;
  int32 value_in_cents = 2;
}

message Product {
  string id = 1;
  int32 price_in_cents = 2;
  string title = 3;
  string description = 4;
  Discount discount = 5;

}

message User {
  string id = 1;
  string first_name = 2;
  string last_name = 3;
  google.protobuf.Timestamp date_of_birth = 4;
}

// The request message contains the product and user identifiers
message DiscountRequest {
  string product_id = 1;
  string user_id = 2;
}

// The response message contains the
message DiscountResponse {
  string product_id = 1;
  int32 full_price_in_cents = 2;
  int32 discounted_price_in_cents = 3;
}

// The request message to obtain a user
message GetUserRequest {
  string user_id = 1;
}

// The request message to obtain a all users
message GetUsersRequest {}

// The request message to obtain a user
message GetUsersResponse {
  repeated User users = 1;
}

// The response message to obtain a user
message GetUserResponse {
  User user = 1;
}

// The request message to obtain a product
message GetProductRequest {
  string product_id = 1;
}

// The request message to obtain a product
message GetProductResponse {
  Product product = 1;
}

// The request message to obtain a product
message GetProductsRequest {}


// The request message to obtain a product
message GetProductsResponse {
  repeated Product products = 1;
}